### ISSUE TITLE
Installing a devcontainer `feature` more than once in a `devcontainer.json`

### ISSUE THREAD

**Issue by joshspicer** [devcontainers/spec#44]
The current [dev container features specification proposal](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md#execution) is based on an array-like syntax for the 'features' attribute, and thus allows for the invocation of a single feature >1 with different options.

An execution goal stated in the spec is that 
> "It should be possible to execute a feature multiple times with different parameters."

We are currently discussing how best to tackle [feature installation order (#43)](https://github.com/devcontainers/spec/issues/43), with the discussion leaning toward proceeding without use of an array, meaning that there is a need to explore other mechanisms for invoking a feature to install >1 times.

## Proposal

The propose extending the [value](https://github.com/devcontainers/cli/blob/main/src/spec-configuration/containerFeaturesConfiguration.ts#L33) to include an array of objects.

```jsonc
{
    "image": "azcr.io/joshspicer/my-crazy-complicated-docker-image",
    "features": {
        "devcontainers/features/node": [
            {
                "version": "16"
            },
            {
                "version": "14"
            }
        ],
    },
}
```

This `devcontainer.json` will invoke the `devcontainers/features/node` feature twice with two sets of options **in order** (first 16, and then 14).

**Comment by Chuxel** [comment]
This seems like a reasonable proposal to me and meets some of the things @edgonmsft has highlighted - @chrmarti ?

**Comment by chrmarti** [comment]
Another option is to extend feature options, such that, they can capture arrays and objects of values. That would allow features to capture feature-global options, the list of versions and version-specific options depending on the tools they install.

**Comment by jkeech** [comment]
My preference is to keep it as simple as possible and not overcomplicate things. I like the proposed array syntax with the full options object inside if we are sticking with an object at the top level for the set of features.

Note that this requires using the same version of the feature for each execution (E.g. `devcontainers/features/node@2` would be used for both executions). Do we need to support installing with one feature version with one set of options and another feature version with another set of options? Does the current spec allow specifying the same feature with different versions in the key with the current object syntax, such as the following?

```json
{
    "features": {
        "devcontainers/features/node@1": {
            "version": "14"
        },
        "devcontainers/features/node@2": {
            "version": "16",
            "someNewOptionInV2": "..."
        }
    }
}
```

**Comment by joshspicer** [comment]
@chrmarti  - are you suggesting features that look like this?

```jsonc
"features": {
   "devcontainers/features/node@v0.0.1" : {
       "version": [16, 14, latest]
   }
```

or equivalently with the shorthand that implies "version"

```jsonc
"features": {
   "devcontainers/features/node@v0.0.1" : [16, 14, latest]
```

How would a user modify >1 option for a given feature. I could imagine wanting a user to install a feature >1 time that uses any permutation of options.

**Comment by joshspicer** [comment]
> Does the current spec allow specifying the same feature with different versions in the key with the current object syntax, such as the following?

Yes, given the object syntax and these are two unique keys, I think we should (and would have to) support a declaration like this.

**Comment by edgonmsft** [comment]
I also the array syntax so that the whole options object can be modified for each run. 

The version question is an interesting one. The object syntax does allow us to use different versions since they would have different ids, although ordering will probably get interesting and the user would probably need to override it.

**Comment by joshspicer** [comment]
For the `dotnet` feature, I can imagine users wanting to install `dotnet` twice with the following options.

These are the existing options today, and two plausible ways a user could configure several installations of the feature.

```
{
  "version": 3.1,
  "runtimeOnly": true,
  "installUsingApt": true
}
```

```
{
  "version": 6.0,
  "runtimeOnly": false,
  "installUsingApt": false
}
```

![image](https://user-images.githubusercontent.com/23246594/176004895-6f109ac6-cb79-4c7a-8bf0-dbe55c8e85c3.png)

**Comment by Chuxel** [comment]
@joshspicer @edgonmsft @jkeech @chrmarti

Yeah, to that point - Personally I think we're talking about two separate proposals here.

1. I think that we will need an array of strings option type.  Right now if we added the ability to install a list of packages for a feature like homebrew, you'd need to do a space separated list which isn't great. So an array of strings is a reasonable type.
2. This proposal is going beyond those basic needs and trying to meet the need of varying sets of configuration options for a given feature.  You can't resolve that with just an array option type. Something along these lines will likely be needed as well. That is what Josh is showing in https://github.com/devcontainers/spec/issues/44#issuecomment-1167683951.

Let's not discuss this as an "or", but rather whether we think this is needed now verses an array option. An alternative is also that we implement both at once.

**Comment by joshspicer** [comment]
@Chuxel I agree we may be getting side-tracked.  I added a new issue to address (1) in your comment: https://github.com/devcontainers/spec/issues/57

I want this proposal to focus on (2)

**Comment by joshspicer** [comment]
This is the current interface for the `python` feature

![image](https://user-images.githubusercontent.com/23246594/176009859-bff5c849-3e8f-4e24-a8a9-541bb22f5920.png)

**Comment by jkeech** [comment]
@chrmarti how would complex objects get passed in and parsed by the install script? I think it will end up being more complicated for both end users and feature authors to deal with complex option types and instead that we should support executing a feature multiple times by looping on the outside with different sets of simple options.

**Comment by chrmarti** [comment]
This seems simple for the end-user. For the feature we can continue to pass in a flattened list of attributes (possibly omitting the nested ones) and also include a variable with the full options JSON. I expect most features to require only simple options.

**Comment by alexdima** [comment]
We agree:
* it is our goal that features are idempotent / re-entrant / don't explode if they are executed multiple times and we operate under this default.
* [ ] a feature should be able to declare in its manifest if it is safe to be executed multiple times or not
* [ ] when running the "Add Development Container Configuration Files..." command, the features already installed in the base image should be indicated in the quick pick and potentially not suggested via ctrl+space.
* [ ] the feature test command could ensure features don't explode if they are executed multiple times.

**Comment by joshspicer** [comment]
@Alegrowin There's no workaround within the spec today to install the same Feature twice.  We're exploring changing this soon with the introduction of composite/dependencies, but this exact functionality in on our backlog (although I agree it's functionality that would be very nice to have).  

Today, a lot of the Features in the wild wouldn't behave nicely if installed twice (either short-circuit early or overwrite existing installs).  I personally don't want to introduce a solution the right guidance in place for authors to make the behavior predictable.

For this example, I think a Feature that proxies to another plugin manager should naturally accept >1 plugin. For example, I think it would be nice if upstream the feature worked like so:

```
{
    "features": {
		"ghcr.io/devcontainers-contrib/features/asdf-package:1": {
			"plugin": "aws-vault,consul"
		}
    }
}
```

With Feature dependencies though, without rigid rules around options we'd still need to support the case you're illustrating above (installing the Feature twice). And I also understand that someone may want to use a Feature as-is, even if it's a bit ugly.  

Thanks for putting this back on our radar as useful to you.  Taking this feedback back to our team.

**Comment by endquote** [comment]
> Meanwhile, is there any workaround for something like this?
> 
> ```json
> {
>     "features": {
> 		"ghcr.io/devcontainers-contrib/features/asdf-package:1": {
> 			"plugin": "aws-vault"
> 		}
>        		"ghcr.io/devcontainers-contrib/features/asdf-package:1": {
> 			"plugin": "consul"
> 		}
>     }
> }
> ```

exactly what I was trying to do

**Comment by Xcopz1** [comment]
{
    "features": {
		"ghcr.io/devcontainers-contrib/features/asdf-package:1": {
			"plugin": "aws-vault"
		}
       		"ghcr.io/devcontainers-contrib/features/asdf-package:1": {
			"plugin": "consul"
		}
    }
}

**Comment by Xcopz1** [comment]
{
    "Futures":     {
			“ghcr.io/devcontainers-contrib/features/asdf-package:1” :   {                              “Plug-in”: “aws-vault”
		}
       "ghcr.io/devcontainers-contrib/features/asdf-package:1" :   {
			“plug-in”: “consul”
		}
    }
}

**Comment by PmPPolska** [comment]
•••

⤴️


wt., 26 lis 2024, 15:40 użytkownik Xcopz1 ***@***.***>
napisał:

**Comment by djmcgreal-cc** [comment]
There are lots of package manager features available but few seem to implement their own iterations. Seems like it's pretty soon before running into this and having to just go back to managing one's own container build.

**Comment by Alegrowin** [comment]
Update: I've published yesterday a devcontainer feature that solve for the asdf specific issue. I've been using it for a while now. 

Users can `opt-in` to `automatically` add plugins, install and set the tools versions from `.tool-versions`

https://github.com/Alegrowin/devcontainer-features/
